File: src/opencode/storage.ts
/**
 * OpenCode 本地数据仓 — data/opencode/ 分层持久化与增量同步
 * ponytail: 文件级同步无锁，单进程串行写已足够；如需并发需加文件锁
 */
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";
import type { OpenCodeCostsResult, OpenCodeUsageRecord } from "./types.ts";

export interface SyncOptions {
  workspaceId?: string;
  full?: boolean;
  limit?: number;
}

export interface SyncResult {
  added: number;
  pages: number;
  elapsedMs: number;
  lastSyncedTime: string | null;
}

export interface HistoryFilter {
  model?: string;
  sessionId?: string;
  sessionID?: string; // alias
  since?: string;
  until?: string;
  // also support time range via desde/hasta strings
}

interface HistoryFile {
  records: OpenCodeUsageRecord[];
  lastSyncedTime: string | null;
  updatedAt: string;
}

type CostsFile = Record<string, OpenCodeCostsResult>;

const CSV_FIELDS: (keyof OpenCodeUsageRecord)[] = [
  "id",
  "workspaceID",
  "timeCreated",
  "timeUpdated",
  "timeDeleted",
  "model",
  "provider",
  "inputTokens",
  "outputTokens",
  "reasoningTokens",
  "cacheReadTokens",
  "cacheWrite5mTokens",
  "cacheWrite1hTokens",
  "cost",
  "keyID",
  "sessionID",
  "enrichment",
];

function csvEscape(v: unknown): string {
  if (v === null || v === undefined) return "";
  let s: string;
  if (typeof v === "string") s = v;
  else if (typeof v === "number" || typeof v === "boolean") s = String(v);
  else s = JSON.stringify(v);
  if (s.includes('"') || s.includes(",") || s.includes("\n") || s.includes("\r")) {
    return `"${s.replace(/"/g, '""')}"`;
  }
  return s;
}

function keyFor(year: number, month: number): string {
  return `${year}-${String(month).padStart(2, "0")}`;
}

function sortDesc(a: OpenCodeUsageRecord, b: OpenCodeUsageRecord): number {
  const ta = Date.parse(a.timeCreated);
  const tb = Date.parse(b.timeCreated);
  if (Number.isNaN(ta) && Number.isNaN(tb)) return 0;
  if (Number.isNaN(ta)) return 1;
  if (Number.isNaN(tb)) return -1;
  return tb - ta; // 逆序
}

export class OpenCodeStorage {
  readonly dataDir: string;

  constructor(dataDir: string = "data/opencode") {
    this.dataDir = dataDir;
  }

  private costsPath(): string {
    return join(this.dataDir, "costs.json");
  }
  private historyPath(): string {
    return join(this.dataDir, "history.json");
  }
  private csvPath(): string {
    return join(this.dataDir, "history.csv");
  }

  // T1
  async ensureDataDir(): Promise<void> {
    await mkdir(this.dataDir, { recursive: true });
    // costs.json
    try {
      await readFile(this.costsPath(), "utf-8");
    } catch {
      await writeFile(this.costsPath(), JSON.stringify({}, null, 2), "utf-8");
    }
    // history.json
    try {
      await readFile(this.historyPath(), "utf-8");
    } catch {
      const init: HistoryFile = { records: [], lastSyncedTime: null, updatedAt: new Date().toISOString() };
      await writeFile(this.historyPath(), JSON.stringify(init, null, 2), "utf-8");
    }
  }

  private async readCostsFile(): Promise<CostsFile> {
    try {
      const txt = await readFile(this.costsPath(), "utf-8");
      const parsed = JSON.parse(txt) as unknown;
      if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
        // support both { "2026-08": result } and { entries: {...} }
        const obj = parsed as Record<string, unknown>;
        if ("entries" in obj && obj.entries && typeof obj.entries === "object" && !Array.isArray(obj.entries)) {
          return obj.entries as CostsFile;
        }
        return parsed as CostsFile;
      }
      return {};
    } catch {
      return {};
    }
  }

  private async writeCostsFile(data: CostsFile): Promise<void> {
    await mkdir(this.dataDir, { recursive: true });
    await writeFile(this.costsPath(), JSON.stringify(data, null, 2), "utf-8");
  }

  private async readHistoryFile(): Promise<HistoryFile> {
    try {
      const txt = await readFile(this.historyPath(), "utf-8");
      const parsed = JSON.parse(txt) as unknown;
      if (parsed && typeof parsed === "object" && !Array.isArray(parsed)) {
        const obj = parsed as Record<string, unknown>;
        const records = Array.isArray(obj.records) ? (obj.records as OpenCodeUsageRecord[]) : [];
        const lastSyncedTime = typeof obj.lastSyncedTime === "string" ? (obj.lastSyncedTime as string) : null;
        const updatedAt = typeof obj.updatedAt === "string" ? (obj.updatedAt as string) : new Date().toISOString();
        return { records, lastSyncedTime, updatedAt };
      }
      return { records: [], lastSyncedTime: null, updatedAt: new Date().toISOString() };
    } catch {
      return { records: [], lastSyncedTime: null, updatedAt: new Date().toISOString() };
    }
  }

  private async writeHistoryFile(data: HistoryFile): Promise<void> {
    await mkdir(this.dataDir, { recursive: true });
    await writeFile(this.historyPath(), JSON.stringify(data, null, 2), "utf-8");
  }

  // T2
  async saveCosts(year: number, month: number, result: OpenCodeCostsResult): Promise<void> {
    await this.ensureDataDir();
    const costs = await this.readCostsFile();
    costs[keyFor(year, month)] = result;
    await this.writeCostsFile(costs);
  }

  async getCosts(year: number, month: number): Promise<OpenCodeCostsResult | null> {
    const costs = await this.readCostsFile();
    const k = keyFor(year, month);
    return costs[k] ?? null;
  }

  async listCosts(): Promise<Array<{ year: number; month: number; result: OpenCodeCostsResult }>> {
    const costs = await this.readCostsFile();
    const out: Array<{ year: number; month: number; result: OpenCodeCostsResult }> = [];
    for (const [k, v] of Object.entries(costs)) {
      const m = k.match(/^(\d{4})-(\d{1,2})$/);
      if (!m) continue;
      out.push({ year: Number(m[1]), month: Number(m[2]), result: v });
    }
    out.sort((a, b) => a.year !== b.year ? a.year - b.year : a.month - b.month);
    return out;
  }

  // 支持查询全部 costs map（供 04 使用）
  async getAllCosts(): Promise<CostsFile> {
    return this.readCostsFile();
  }

  // T3
  async mergeHistory(records: OpenCodeUsageRecord[]): Promise<{ added: number; total: number }> {
    await this.ensureDataDir();
    const file = await this.readHistoryFile();
    const existing = file.records;
    const existingMap = new Map<string, OpenCodeUsageRecord>();
    for (const r of existing) existingMap.set(r.id, r);

    let added = 0;
    // 去重：保留最新（后到的覆盖）
    for (const r of records) {
      if (!existingMap.has(r.id)) added++;
      existingMap.set(r.id, r);
    }

    const merged = Array.from(existingMap.values());
    merged.sort(sortDesc);

    const lastSyncedTime = merged.length > 0 ? merged[0].timeCreated : file.lastSyncedTime;
    // 若 merged 为空保留原游标（可能 null）
    const next: HistoryFile = {
      records: merged,
      lastSyncedTime,
      updatedAt: new Date().toISOString(),
    };
    await this.writeHistoryFile(next);
    // 自动生成 CSV
    await this.exportCsv();
    return { added, total: merged.length };
  }

  // T4
  async exportCsv(filePath?: string): Promise<string> {
    const file = await this.readHistoryFile();
    const records = file.records;
    const header = CSV_FIELDS.join(",");
    const lines: string[] = [header];
    for (const r of records) {
      const row = CSV_FIELDS.map((f) => csvEscape((r as unknown as Record<string, unknown>)[f as string])).join(",");
      lines.push(row);
    }
    const csv = lines.join("\n") + "\n";
    const target = filePath ?? this.csvPath();
    await mkdir(this.dataDir, { recursive: true });
    await writeFile(target, csv, "utf-8");
    return csv;
  }

  // T6
  async loadHistory(): Promise<HistoryFile> {
    await this.ensureDataDir();
    return this.readHistoryFile();
  }

  async getHistory(filter?: HistoryFilter): Promise<OpenCodeUsageRecord[]> {
    const file = await this.readHistoryFile();
    let records = file.records;
    if (!filter) return records;
    const model = filter.model;
    const sessionId = filter.sessionId ?? filter.sessionID;
    const since = filter.since;
    const until = filter.until;

    if (model) records = records.filter((r) => r.model === model);
    if (sessionId) records = records.filter((r) => r.sessionID === sessionId);
    if (since) {
      const s = Date.parse(since);
      if (!Number.isNaN(s)) records = records.filter((r) => Date.parse(r.timeCreated) >= s);
    }
    if (until) {
      const u = Date.parse(until);
      if (!Number.isNaN(u)) records = records.filter((r) => Date.parse(r.timeCreated) <= u);
    }
    return records;
  }

  async clear(): Promise<void> {
    await this.ensureDataDir();
    const empty: HistoryFile = { records: [], lastSyncedTime: null, updatedAt: new Date().toISOString() };
    await this.writeHistoryFile(empty);
    await this.writeCostsFile({});
    // csv 也清空为仅表头
    await this.exportCsv();
  }

  async reset(): Promise<void> {
    return this.clear();
  }

  // T5 sync
  async sync(client: { getUsageInfo: (workspaceId: string, page: number) => Promise<OpenCodeUsageRecord[]> }, opts: SyncOptions): Promise<SyncResult>;
  async sync(workspaceId: string, client: { getUsageInfo: (workspaceId: string, page: number) => Promise<OpenCodeUsageRecord[]> }, opts?: SyncOptions): Promise<SyncResult>;
  async sync(client: { getUsageInfo: (workspaceId: string, page: number) => Promise<OpenCodeUsageRecord[]> }, workspaceId: string, opts?: SyncOptions): Promise<SyncResult>;
  async sync(
    clientOrWorkspaceId: string | { getUsageInfo: (workspaceId: string, page: number) => Promise<OpenCodeUsageRecord[]> },
    workspaceIdOrOpts?: string | SyncOptions | { getUsageInfo: (workspaceId: string, page: number) => Promise<OpenCodeUsageRecord[]> },
    optsMaybe?: SyncOptions,
  ): Promise<SyncResult> {
    let client: { getUsageInfo: (workspaceId: string, page: number) => Promise<OpenCodeUsageRecord[]> };
    let workspaceId: string | undefined;
    let opts: SyncOptions | undefined;

    // overload 1: sync(workspaceId, client, opts)
    if (typeof clientOrWorkspaceId === "string") {
      workspaceId = clientOrWorkspaceId;
      client = workspaceIdOrOpts as unknown as typeof client;
      opts = optsMaybe;
    } else {
      client = clientOrWorkspaceId;
      if (typeof workspaceIdOrOpts === "string") {
        workspaceId = workspaceIdOrOpts;
        opts = optsMaybe;
      } else {
        opts = workspaceIdOrOpts as SyncOptions | undefined;
        workspaceId = opts?.workspaceId;
      }
    }

    if (!workspaceId) throw new Error("sync 需要 workspaceId（传 workspaceId 参数或 opts.workspaceId）");
    if (!client || typeof client.getUsageInfo !== "function") throw new Error("sync 需要提供含 getUsageInfo 的 client");

    const full = opts?.full ?? false;
    const limit = opts?.limit;

    await this.ensureDataDir();
    const file = await this.readHistoryFile();
    const existingIds = new Set(file.records.map((r) => r.id));
    const lastSyncedTime = file.lastSyncedTime;
    const lastTs = lastSyncedTime ? Date.parse(lastSyncedTime) : null;

    const collected: OpenCodeUsageRecord[] = [];
    let pages = 0;
    const start = Date.now();
    // eslint-disable-next-line no-constant-condition
    for (let page = 0; true; page++) {
      if (limit !== undefined && pages >= limit) break;
      const batch = await client.getUsageInfo(workspaceId, page);
      pages++;
      if (!batch || batch.length === 0) break;

      if (full) {
        collected.push(...batch);
        continue;
      }

      // 检查截断点
      let cutoffIdx = -1;
      for (let i = 0; i < batch.length; i++) {
        const r = batch[i];
        if (existingIds.has(r.id)) {
          cutoffIdx = i;
          break;
        }
        if (lastTs !== null) {
          const t = Date.parse(r.timeCreated);
          if (!Number.isNaN(t) && t <= lastTs) {
            cutoffIdx = i;
            break;
          }
        }
      }
      if (cutoffIdx !== -1) {
        // 只收集截断点之前的
        collected.push(...batch.slice(0, cutoffIdx));
        break;
      } else {
        collected.push(...batch);
      }
      // 若本页已出现截断或空，已 break；否则继续下一页
      // 为了避免无限循环，若未截断但 batch 为空已在顶部 break
    }

    let added = 0;
    if (collected.length > 0) {
      const res = await this.mergeHistory(collected);
      added = res.added;
    } else {
      // 即使无新增，也需刷新 updatedAt？保持 lastSyncedTime 不变，但若全空且原无记录则仍 null
      // 不额外写文件以免无意义抖动，但 sync 返回需读最新游标
    }

    const after = await this.readHistoryFile();
    const elapsedMs = Date.now() - start;
    return { added, pages, elapsedMs, lastSyncedTime: after.lastSyncedTime };
  }
}

File: opencode-analyzer/internal/opencode/storage.go
package opencode

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

type HistoryFile struct {
	Records        []UsageRecord `json:"records"`
	LastSyncedTime *string       `json:"lastSyncedTime"`
	UpdatedAt      string        `json:"updatedAt"`
}

type Storage struct {
	dataDir string
}

func NewStorage(dataDir string) *Storage {
	if strings.TrimSpace(dataDir) == "" {
		dataDir = "data/opencode"
	}
	return &Storage{dataDir: dataDir}
}

func (s *Storage) DataDir() string { return s.dataDir }

func (s *Storage) EnsureDataDir() error {
	if err := os.MkdirAll(s.dataDir, 0755); err != nil {
		return err
	}
	if _, err := os.Stat(s.costsPath()); os.IsNotExist(err) {
		if err := os.WriteFile(s.costsPath(), []byte("{}\n"), 0644); err != nil {
			return err
		}
	}
	if _, err := os.Stat(s.historyPath()); os.IsNotExist(err) {
		history := HistoryFile{Records: []UsageRecord{}, UpdatedAt: time.Now().UTC().Format(time.RFC3339Nano)}
		data, err := json.MarshalIndent(history, "", "  ")
		if err != nil {
			return err
		}
		if err := os.WriteFile(s.historyPath(), append(data, '\n'), 0644); err != nil {
			return err
		}
	}
	return nil
}

func (s *Storage) costsPath() string   { return filepath.Join(s.dataDir, "costs.json") }
func (s *Storage) historyPath() string { return filepath.Join(s.dataDir, "history.json") }
func (s *Storage) csvPath() string     { return filepath.Join(s.dataDir, "history.csv") }
func (s *Storage) lockPath() string    { return filepath.Join(s.dataDir, ".lock") }

// Lock acquires a cross-process exclusive lock for sync writes.
func (s *Storage) Lock() (func(), error) {
	if err := s.EnsureDataDir(); err != nil {
		return nil, err
	}
	return lockFile(s.lockPath())
}

func (s *Storage) readCosts() (map[string]CostsResult, error) {
	data, err := os.ReadFile(s.costsPath())
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]CostsResult{}, nil
		}
		return nil, err
	}
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, err
	}
	if entries, ok := raw["entries"]; ok {
		var nested map[string]CostsResult
		if err := json.Unmarshal(entries, &nested); err == nil {
			return nested, nil
		}
	}
	out := make(map[string]CostsResult)
	for key, value := range raw {
		var result CostsResult
		if err := json.Unmarshal(value, &result); err == nil {
			out[key] = result
		}
	}
	return out, nil
}

func (s *Storage) GetCosts(year, month int) (*CostsResult, error) {
	costs, err := s.readCosts()
	if err != nil {
		return nil, err
	}
	result, ok := costs[fmt.Sprintf("%04d-%02d", year, month)]
	if !ok {
		return nil, nil
	}
	return &result, nil
}

func (s *Storage) SaveCosts(year, month int, result CostsResult) error {
	if err := s.EnsureDataDir(); err != nil {
		return err
	}
	costs, err := s.readCosts()
	if err != nil {
		return err
	}
	costs[fmt.Sprintf("%04d-%02d", year, month)] = result
	data, err := json.MarshalIndent(costs, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.costsPath(), append(data, '\n'), 0644)
}

func (s *Storage) ListCosts() ([]struct {
	Year   int
	Month  int
	Result CostsResult
}, error) {
	costs, err := s.readCosts()
	if err != nil {
		return nil, err
	}
	out := make([]struct {
		Year   int
		Month  int
		Result CostsResult
	}, 0, len(costs))
	for key, result := range costs {
		var year, month int
		if _, err := fmt.Sscanf(key, "%d-%d", &year, &month); err != nil || month < 1 || month > 12 {
			continue
		}
		out = append(out, struct {
			Year   int
			Month  int
			Result CostsResult
		}{Year: year, Month: month, Result: result})
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Year != out[j].Year {
			return out[i].Year < out[j].Year
		}
		return out[i].Month < out[j].Month
	})
	return out, nil
}

func (s *Storage) LoadHistory() (*HistoryFile, error) {
	data, err := os.ReadFile(s.historyPath())
	if err != nil {
		if os.IsNotExist(err) {
			return &HistoryFile{Records: []UsageRecord{}}, nil
		}
		return nil, err
	}
	var history HistoryFile
	if err := json.Unmarshal(data, &history); err != nil {
		return nil, err
	}
	if history.Records == nil {
		history.Records = []UsageRecord{}
	}
	if history.UpdatedAt == "" {
		history.UpdatedAt = time.Now().UTC().Format(time.RFC3339Nano)
	}
	return &history, nil
}

func (s *Storage) SaveHistory(records []UsageRecord, lastSyncedTime *string) error {
	if err := s.EnsureDataDir(); err != nil {
		return err
	}
	ordered := append([]UsageRecord(nil), records...)
	sortUsage(ordered)
	history := HistoryFile{
		Records:        ordered,
		LastSyncedTime: lastSyncedTime,
		UpdatedAt:      time.Now().UTC().Format(time.RFC3339Nano),
	}
	data, err := json.MarshalIndent(history, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(s.historyPath(), append(data, '\n'), 0644); err != nil {
		return err
	}
	return s.ExportCSV(ordered)
}

func (s *Storage) MergeHistory(records []UsageRecord) (int, error) {
	if err := s.EnsureDataDir(); err != nil {
		return 0, err
	}
	history, err := s.LoadHistory()
	if err != nil {
		return 0, err
	}
	byID := make(map[string]UsageRecord, len(history.Records)+len(records))
	for _, record := range history.Records {
		byID[record.ID] = record
	}
	added := 0
	for _, record := range records {
		if _, exists := byID[record.ID]; !exists {
			added++
		}
		byID[record.ID] = record
	}
	merged := make([]UsageRecord, 0, len(byID))
	for _, record := range byID {
		merged = append(merged, record)
	}
	sortUsage(merged)
	var last *string
	if len(merged) > 0 {
		value := merged[0].TimeCreated
		last = &value
	} else {
		last = history.LastSyncedTime
	}
	if err := s.SaveHistory(merged, last); err != nil {
		return 0, err
	}
	return added, nil
}

func sortUsage(records []UsageRecord) {
	sort.SliceStable(records, func(i, j int) bool {
		left, leftOK := parseTimestamp(records[i].TimeCreated)
		right, rightOK := parseTimestamp(records[j].TimeCreated)
		if leftOK && rightOK && !left.Equal(right) {
			return left.After(right)
		}
		return records[i].TimeCreated > records[j].TimeCreated
	})
}

func EncodeCSV(records []UsageRecord) ([]byte, error) {
	var buffer bytes.Buffer
	writer := csv.NewWriter(&buffer)
	headers := []string{
		"id", "workspaceID", "timeCreated", "timeUpdated", "timeDeleted",
		"model", "provider", "inputTokens", "outputTokens", "reasoningTokens",
		"cacheReadTokens", "cacheWrite5mTokens", "cacheWrite1hTokens", "cost",
		"keyID", "sessionID", "enrichment",
	}
	if err := writer.Write(headers); err != nil {
		return nil, err
	}
	for _, record := range records {
		deleted, session := "", ""
		if record.TimeDeleted != nil {
			deleted = *record.TimeDeleted
		}
		if record.SessionID != nil {
			session = *record.SessionID
		}
		enrichment := ""
		if record.Enrichment != nil {
			data, err := json.Marshal(record.Enrichment)
			if err != nil {
				return nil, err
			}
			enrichment = string(data)
		}
		row := []string{
			record.ID,
			record.WorkspaceID,
			record.TimeCreated,
			record.TimeUpdated,
			deleted,
			record.Model,
			record.Provider,
			formatNumber(record.InputTokens),
			formatNumber(record.OutputTokens),
			formatNumber(record.ReasoningTokens),
			formatNumber(record.CacheReadTokens),
			formatNumber(record.CacheWrite5mTokens),
			formatNumber(record.CacheWrite1hTokens),
			formatNumber(record.Cost),
			record.KeyID,
			session,
			enrichment,
		}
		if err := writer.Write(row); err != nil {
			return nil, err
		}
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		return nil, err
	}
	return buffer.Bytes(), nil
}

func (s *Storage) ExportCSV(records []UsageRecord) error {
	if err := s.EnsureDataDir(); err != nil {
		return err
	}
	data, err := EncodeCSV(records)
	if err != nil {
		return err
	}
	return os.WriteFile(s.csvPath(), data, 0644)
}

func (s *Storage) GetHistory(filter HistoryFilter, page, size int) ([]UsageRecord, int, error) {
	history, err := s.LoadHistory()
	if err != nil {
		return nil, 0, err
	}
	filtered := make([]UsageRecord, 0, len(history.Records))
	for _, record := range history.Records {
		if filter.Model != "" && record.Model != filter.Model {
			continue
		}
		if filter.SessionID != "" && (record.SessionID == nil || *record.SessionID != filter.SessionID) {
			continue
		}
		if filter.Since != "" {
			if timestamp, ok := parseTimestamp(record.TimeCreated); ok {
				if since, valid := parseTimestamp(filter.Since); valid && timestamp.Before(since) {
					continue
				}
			}
		}
		if filter.Until != "" {
			if timestamp, ok := parseTimestamp(record.TimeCreated); ok {
				if until, valid := parseTimestamp(filter.Until); valid && timestamp.After(until) {
					continue
				}
			}
		}
		filtered = append(filtered, record)
	}
	total := len(filtered)
	if page <= 0 || size <= 0 {
		return filtered, total, nil
	}
	start := (page - 1) * size
	if start > total {
		start = total
	}
	end := start + size
	if end > total {
		end = total
	}
	return filtered[start:end], total, nil
}

func (s *Storage) Clear() error {
	if err := s.EnsureDataDir(); err != nil {
		return err
	}
	if err := s.SaveHistory([]UsageRecord{}, nil); err != nil {
		return err
	}
	return os.WriteFile(s.costsPath(), []byte("{}\n"), 0644)
}

func (s *Storage) Reset() error { return s.Clear() }

type UsageClient interface {
	GetUsageHistory(workspaceID string, page int) ([]UsageRecord, error)
}

func (s *Storage) Sync(client UsageClient, options SyncOptions) (SyncResult, error) {
	if strings.TrimSpace(options.WorkspaceID) == "" {
		return SyncResult{}, fmt.Errorf("sync 需要 workspaceId")
	}
	if client == nil {
		return SyncResult{}, fmt.Errorf("sync 需要提供 OpenCode client")
	}
	if options.Limit < 0 {
		return SyncResult{}, fmt.Errorf("无效 limit: %d（需为非负整数）", options.Limit)
	}
	if err := s.EnsureDataDir(); err != nil {
		return SyncResult{}, err
	}
	history, err := s.LoadHistory()
	if err != nil {
		return SyncResult{}, err
	}
	existing := make(map[string]bool, len(history.Records))
	for _, record := range history.Records {
		existing[record.ID] = true
	}
	last, lastOK := parseTimestamp(pointerValue(history.LastSyncedTime))
	collected := make([]UsageRecord, 0)
	pages := 0
	maxPages := options.Limit
	if maxPages == 0 {
		// ponytail: finite safety cap for a broken pagination endpoint; raise only if OpenCode history exceeds 200 pages.
		maxPages = 200
	}
	start := time.Now()
	for page := 0; pages < maxPages; page++ {
		batch, err := client.GetUsageHistory(options.WorkspaceID, page)
		pages++
		if err != nil {
			return SyncResult{}, err
		}
		if len(batch) == 0 {
			break
		}
		if options.Full {
			collected = append(collected, batch...)
			continue
		}
		cutoff := -1
		for i, record := range batch {
			if existing[record.ID] {
				cutoff = i
				break
			}
			if lastOK {
				if timestamp, ok := parseTimestamp(record.TimeCreated); ok && !timestamp.After(last) {
					cutoff = i
					break
				}
			}
		}
		if cutoff >= 0 {
			collected = append(collected, batch[:cutoff]...)
			break
		}
		collected = append(collected, batch...)
	}
	added := 0
	if len(collected) > 0 {
		added, err = s.MergeHistory(collected)
	} else {
		err = s.ExportCSV(history.Records)
	}
	if err != nil {
		return SyncResult{}, err
	}
	after, err := s.LoadHistory()
	if err != nil {
		return SyncResult{}, err
	}
	return SyncResult{
		Added:          added,
		Pages:          pages,
		ElapsedMs:      time.Since(start).Milliseconds(),
		LastSyncedTime: pointerValue(after.LastSyncedTime),
	}, nil
}

func pointerValue(value *string) string {
	if value == nil {
		return ""
	}
	return *value
}

func parseTimestamp(value string) (time.Time, bool) {
	if value == "" {
		return time.Time{}, false
	}
	parsed, err := time.Parse(time.RFC3339Nano, value)
	return parsed, err == nil
}

func formatNumber(value float64) string {
	return strconv.FormatFloat(value, 'f', -1, 64)
}

func CsvEscape(value string) string {
	if strings.ContainsAny(value, "\",\n\r") {
		return `"` + strings.ReplaceAll(value, `"`, `""`) + `"`
	}
	return value
}

func ToFloatStr(value float64) string { return strconv.FormatFloat(value, 'f', -1, 64) }
